### PR TITLE
feat: add new cryptography subject for building basic zkp protocol from scratch

### PR DESCRIPTION
**Pull request by 0xpanoramix**
This PR adds a new subject `Build a basic ZKP protocol from scratch` and is currently blocked by #1.

### CODE REVIEW

**Comment by TomChv** [comment]
I'm waiting for #1 to be merged and adopted before reviewing this PR.

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+1. Create a `verifier.py` file.
+   1. Generate a random number `r`. Use the `random.randint(-127, 128)` method
+   call to sample a random `int8`.
+   2. Evaluate `t = int(target_polynomial(r))` using the target polynomial found in the `common.py`
+   file.
+   3. Print the variable `r`.
+2. In the `prover.py` file:
+   1. After the last line, add an `input()` call which stores the value passed in the standard input
+   in a variable called `r`. Don't forget to surround the input call with an integer cast `int(..)`.
+   2. Compute the value of a variable called `p` equal to the evaluation of the secret polynomial
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:121]:
```suggestion
   2. Compute the value of a variable `p` equal to the evaluation of the secret polynomial
```

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+   using the value `r`.
+   3. Compute the value of a variable called `h` equal to the evaluation of the arbitrary
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:123]:
```suggestion
   3. Compute the value of a variable `h` equal to the evaluation of the arbitrary
```

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+   polynomial using the value `r`.
+   4. Print the values `p` and `h`.
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:125]:
```suggestion
   4. Print the variables `p` and `h`.
```

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+### ✅ Validation
+
+In a first terminal, run the following command:
+```shell
+python3 verifier.py
+```
+
+Which should print (note that the data sampled can differ as it is a random number):
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:134]:
```suggestion
Which should print (note that the data sampled can differ as it depends on a randomly generated number):
```

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+   1. Add two `int(float(input()))` calls and store the result in two variables, namely `p` and `h`.
+   2. Compute `h * t`.
+   3. If the resulting value is equal to `p`, print the following sentence: `Okay, it seems you
+   know all the roots for your third degree polynomial`.
+   4. Otherwise, print: `Mmmmh, I think you don't know all the roots of this third degree
+   polynomial`
+
+### ✅ Validation
+
+Repeat the process from the last step.
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:175]:
```suggestion
Repeat the process from step 2.
```

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+who's waiting for us, you should get:
+```shell
+12
+44
+Mmmmh, I think you don't know all the roots of this third degree polynomial
+```
+
+### 📚 Documentation
+
+Nothing here !
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:201]: Then remove it

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+## Possible Improvements
+
+This is a very basic implementation. You can improve it for example by not having the polynomials
```
Comment (TomChv) [inline cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:205]: What about a secured version? What about doing the same for a 4-degree polynomial?

**Review by TomChv — COMMENTED** [review]
That's nice work! I left some suggestions ;)

Global comments:
- Rename the directory `building_a_simple_zkp_protocol_from_scratch` with its index `0x01` so the future workshops will be sorted.
- Make short sentences that go straight to the point

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md (hunk 7 above)
Comment (0xpanoramix) [reply cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:205]: This is the topic of the next subjects 👀

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md (hunk 6 above)
Comment (0xpanoramix) [reply cryptography/building_a_simple_zkp_protocol_from_scratch/README.md:201]: Just a typo, thks !

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
+# Build a simple ZKP Protocol from scratch
+
+In this workshop, you will create a very simple Zero-Knowledge Proof protocol that can be used by a
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply cryptography/building_a_simple_zkp_protocol_from_scratch/README.md]: Added in 6b961c700511ddb991df0bb610c29f860646a336

File: cryptography/building_a_simple_zkp_protocol_from_scratch/SETUP.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
+# Setup
+
+In this workshop, we are using Python and a virtual environment.
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply cryptography/building_a_simple_zkp_protocol_from_scratch/SETUP.md]: Added in e56e3c6b217a6cb5fd988b679ea730bfb22b1531

File: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md (hunk 6 above)
Comment (0xpanoramix) [reply cryptography/building_a_simple_zkp_protocol_from_scratch/README.md]: Removed in 1aec933fa45548320dbc71bf14c758d45fcb44db

File: cryptography/building_a_simple_zkp_protocol_from_scratch/SETUP.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+Run the following command to activate the virtual environment:
+```shell
+source basic_zkp_protocol_venv/bin/activate
+```
+
+Install numpy using the following command:
+```shell
+python3 -m pip install numpy
+```
+
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply cryptography/building_a_simple_zkp_protocol_from_scratch/SETUP.md]: Added in 3e211c19977c28ab564c1d16271ccceebd9a1036

**Review by TomChv — APPROVED** [review]
LGTM

**Pull request merged**